Remove unsafe usage of resource id

rosstimothy opened the pull request:
The transport used by app service was using a sync.Map keyed by the resource id to store app servers. However, the id of a resource is not guaranteed to be unique which could cause apps to be missed. The sync.Map was removed in favor of a mutex and iterating over the configured servers slice directly.

mdwn commented on lib/web/app/transport.go:
```diff
@@ -231,41 +225,30 @@ func (t *transport) rewriteRequest(r *http.Request) error {
-			}
+func (t *transport) DialContext(ctx context.Context, _, _ string) (conn net.Conn, err error) {
+	servers := t.c.servers[:0]
+	for i, appServer := range t.c.servers {
```
Do we need a mutex here? Feels like we might need to since we modify `t.c.servers` later, unless `DialContext` is thread safe.

mdwn commented on lib/web/app/transport.go:
```diff
@@ -231,41 +225,30 @@ func (t *transport) rewriteRequest(r *http.Request) error {
 
-	if err != nil {
-		return nil, trace.Wrap(err)
+	for i := len(servers); i < len(t.c.servers); i++ {
+		t.c.servers[i] = nil
 	}
```
Could you add a comment explaining this? I don't quite understand what this is doing.

gabrielcorado reviewed (approved):
LGTM.

> However, the id of a resource is not guaranteed to be unique which could cause apps to be missed.

That's good to know.

espadolini commented on lib/web/app/transport.go:
```diff
@@ -231,41 +225,36 @@ func (t *transport) rewriteRequest(r *http.Request) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	servers := t.c.servers[:0]
```
We never add to `t.c.servers`, right? Why not just zero out the head and keep the tail, if we're being fancy like this?

rosstimothy committed:
Remove unsafe usage of resource id

The transport used by app service was using a sync.Map keyed by
the resource id to store app servers. However, the id of a resource
is not guaranteed to be unique which could cause apps to be missed.
The sync.Map was removed in favor of a mutex and iterating over the
configured servers slice directly.
